File: desktop-mend/src/test/java/MainTest.java
import co.samco.mend4.desktop.Main;
import co.samco.mend4.desktop.commands.Command;
import co.samco.mend4.desktop.commands.Help;
import commands.TestBase;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.runners.MockitoJUnitRunner;

import java.util.*;

import static org.hamcrest.CoreMatchers.is;
import static org.mockito.Mockito.*;

@RunWith(MockitoJUnitRunner.class)
public class MainTest extends TestBase {
    private Main main;
    private Main.Runner runner;

    @Captor
    private ArgumentCaptor<List<String>> argsCaptor;

    @Before
    public void setup() {
        super.setup();
        runner = mock(Main.Runner.class);
        main = new Main(strings, log);
    }

    @Test
    public void helpTest() {
        Help help = mock(Help.class);
        when(runner.helpCommand()).thenReturn(help);
        for (String s : Help.HELP_ALIASES) {
            main.run(runner, Arrays.asList(s));
        }
        verify(runner, times(Help.HELP_ALIASES.size())).helpCommand();
        verify(help, times(Help.HELP_ALIASES.size())).executeCommand(argsCaptor.capture());
        Assert.assertTrue(argsCaptor.getValue().size() == 0);
    }

    @Test
    public void noArgs() {
        Command defaultCommand = mock(Command.class);
        when(runner.defaultCommand()).thenReturn(defaultCommand);
        main.run(runner, Collections.emptyList());
        verify(runner).defaultCommand();
        verify(defaultCommand).executeCommand(anyListOf(String.class));
    }

    @Test
    public void runCommandWithArgs() {
        exit.expectSystemExit();
        Command command = mock(Command.class);
        runCommand(Arrays.asList("hi"), command);
    }

    @Test
    public void runCommand() {
        exit.expectSystemExit();
        Command command = mock(Command.class);
        runCommand(Collections.emptyList(), command);
    }

    @Test
    public void commandFailed() {
        exit.expectSystemExitWithStatus(-1);
        Command command = mock(Command.class);
        when(command.getExecutionResult()).thenReturn(-1);
        runCommand(Collections.emptyList(), command);
    }

    private void runCommand(List<String> subCommandArgs, Command command) {
        String commandName = "command";
        when(command.isCommandForString(commandName)).thenReturn(true);
        Set<Command> commands = new HashSet<>();
        commands.add(command);
        when(runner.commands()).thenReturn(commands);
        List<String> args = new ArrayList<>();
        args.add(commandName);
        args.addAll(subCommandArgs);
        main.run(runner, args);
        verify(runner).commands();
        verify(command).isCommandForString(commandName);
        verify(command).executeCommand(argsCaptor.capture());
        Assert.assertThat(argsCaptor.getValue(), is(subCommandArgs));
    }
}

File: desktop-mend/src/test/java/MainTest.kt
package co.samco.mend4.desktop

import co.samco.mend4.desktop.commands.Command
import co.samco.mend4.desktop.output.ExitManager
import commands.TestBase
import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Before
import org.junit.Test
import com.nhaarman.mockitokotlin2.KArgumentCaptor
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.argumentCaptor
import com.nhaarman.mockitokotlin2.eq
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.whenever

class MainTest : TestBase() {
    private lateinit var main: Main
    private lateinit var commandRunnerComponent: CommandRunnerComponent
    private lateinit var argsCaptor: KArgumentCaptor<List<String>>

    private val exitManager: ExitManager = mock()

    @Before
    override fun setup() {
        super.setup()
        commandRunnerComponent = mock()
        argsCaptor = argumentCaptor()
        main = Main(
            strings = strings,
            log = log,
            exitManager = exitManager
        )
    }

    private fun commandMock(): Command = mock<Command>().apply {
        whenever(this.commandAliases).thenReturn(listOf("test"))
        whenever(this.usageText).thenReturn("usage")
        whenever(this.descriptionText).thenReturn("description")
    }

    @Test
    fun noArgs() {
        val defaultCommand: Command = commandMock()
        whenever(commandRunnerComponent.defaultCommand()).thenReturn(defaultCommand)
        main.run(commandRunnerComponent, emptyList())
        verify(commandRunnerComponent).defaultCommand()
        verify(defaultCommand).executeCommand(any())
    }

    @Test
    fun runCommandWithArgs() {
        val command: Command = commandMock()
        runCommand(listOf("hi"), command)
        verify(exitManager).exit(eq(0))
    }

    @Test
    fun runCommand() {
        val command: Command = commandMock()
        runCommand(emptyList(), command)
        verify(exitManager).exit(eq(0))
    }

    @Test
    fun commandFailed() {
        val command: Command = commandMock()
        whenever(command.executionResult).thenReturn(-1)
        runCommand(emptyList(), command)
        verify(exitManager).exit(eq(-1))
    }

    private fun runCommand(subCommandArgs: List<String>, command: Command) {
        val commandName = "command"
        whenever(command.isCommandForString(commandName)).thenReturn(true)
        val commands: MutableSet<Command> = HashSet()
        commands.add(command)
        whenever(commandRunnerComponent.commands()).thenReturn(commands)

        val args: MutableList<String> = ArrayList()
        args.add(commandName)
        args.addAll(subCommandArgs)
        main.run(commandRunnerComponent, args)

        verify(commandRunnerComponent).commands()
        verify(command).isCommandForString(commandName)
        verify(command).executeCommand(argsCaptor.capture())
        assertThat(argsCaptor.firstValue, equalTo(subCommandArgs))
    }
}